cmd/thanos/store.go

```diff
@@ -102,35 +100,11 @@ func runStore(
 	component string,
 ) error {
 	{
-		var (
-			bkt objstore.Bucket
-			// closeFn gets called when the sync loop ends to close clients, clean up, etc
-			closeFn = func() error { return nil }
-			bucket  string
-		)
-		if gcsBucket != "" {
-			gcsClient, err := storage.NewClient(context.Background())
-			if err != nil {
-				return errors.Wrap(err, "create GCS client")
-			}
-
-			bkt = gcs.NewBucket(gcsBucket, gcsClient.Bucket(gcsBucket), reg)
-			closeFn = gcsClient.Close
-			bucket = gcsBucket
-		} else if s3Config.Validate() == nil {
-			b, err := s3.NewBucket(s3Config, reg, component)
-			if err != nil {
-				return errors.Wrap(err, "create s3 client")
-			}
-
-			bkt = b
-			bucket = s3Config.Bucket
-		} else {
-			return errors.New("no valid GCS or S3 configuration supplied")
+		bkt, closeFn, err := client.NewBucket(&gcsBucket, *s3Config, reg, component)
```

Comment:
there is yet another issue with this, but actually it was same before (not your fault) -> we need to put additional `closeFn()` for all cases when return err is invoked before we actually start `g.Run()` ): 

But there is some way to make it smarter like this: https://play.golang.org/p/de9RjuF8WxC let me know if it's clear what I mean

Reply:
Hmm, I can't really figure out how to implement this properly in this case :/

Reply:
No worries. Basically, all places where we create `NewBucket` but we don't `defer close()` immdiately we need to ensure we close Client on every flow. Currently we close it only when `run.Group` is actually started. But if this function here `runStore` will return err, close will be never invoked.

So we have 2 options:
1. smart defer
(link from playground)
- return `(err error)` instead of just `error`
- put smart defer:
```
defer func() {
		if err != nil {
			closeFn()
		}
}()
```
2. Carefully lookup all the error case and add `closeFn()` (prone to erros)